php/php.project/src/org/netbeans/modules/php/project/ComputeTestMethodAnnotations.java

```diff
@@ -0,0 +1,189 @@
+            EditorRegistry.removePropertyChangeListener(this);
+        }
+    }
+
+    @Override
+    public void propertyChange(PropertyChangeEvent event) {
+        JTextComponent textComponent = EditorRegistry.lastFocusedComponent();
+        if (textComponent != null) {
+            Document document = textComponent.getDocument();
+            if (document != null) {
+                FileObject fileObject = NbEditorUtilities.getFileObject(document);
+                if (fileObject != null) {
+                    if (FileUtils.isPhpFile(fileObject)) {
+                        if (event.getPropertyName().equals(EditorRegistry.FOCUS_GAINED_PROPERTY)) {
+                            handleFileChange(document);
+                            document.addDocumentListener(ComputeTestMethodAnnotations.this);
+                        } else if (event.getPropertyName().equals(EditorRegistry.FOCUS_LOST_PROPERTY)) {
+                            document.removeDocumentListener(this);
+                        }
+                    }
+                }
+            }
+        }
+    }
+
+    @Override
+    public void insertUpdate(DocumentEvent event) {
+        handleFileChange(event.getDocument());
+    }
+
+    @Override
+    public void removeUpdate(DocumentEvent event) {
+        handleFileChange(event.getDocument());
+    }
+
+    @Override
+    public void changedUpdate(DocumentEvent event) {
+        handleFileChange(event.getDocument());
+    }
+
+    private void handleFileChange(Document doc) {
+        handledDocument = doc;
+        task.schedule(500);
+    }
+
+    @Override
+    public void run() {
+        List<TestMethod> testMethods = new ArrayList<>();
+
+        List<TestMethod> methods = getTestMethods(handledDocument);
+        testMethods.addAll(methods);
+
+        /*
+         * Apparently, this method should update the list of annotations at each call of this method,
+         * when the passed collection of methods changes.
+         * Вut I didn't manage to achieve correct work in this case.
+         * So I made the method call first with the empty collection, to clear the annotation list,
+         * then already with the method collection.
+         */
+        TestMethodController.setTestMethods(handledDocument, Collections.emptyList());
+        TestMethodController.setTestMethods(handledDocument, testMethods);
+    }
+
+    public List<TestMethod> getTestMethods(Document document) {
+        List<TestMethod> testMethods = new ArrayList<>();
+        FileObject fileObject = NbEditorUtilities.getFileObject(document);
+        if (fileObject != null) {
+            PhpProject project = PhpProjectUtils.getPhpProject(fileObject);
+            assert project != null;
+            PhpModule phpModule = project.getPhpModule();
+            for (PhpTestingProvider testingProvider : project.getTestingProviders()) {
+                if (testingProvider.isTestFile(phpModule, fileObject)) {
```

Comment:
Nitpick: inside all loops, consider comparing `document` with `this.document`; if they are different, quite immediately, since the focused document has already changed.

Reply:
Did it, but I'm not sure if it's right.

Reply:
Why do you think so?

Reply:
I'm just not very good at java. ))